File: backend/v2/query/master_data.py
"""V2 public master data API backed by Aurora PostgreSQL through RDS Data API."""

from __future__ import annotations

import json
import logging
import os
from dataclasses import dataclass
from typing import Any, Iterable

LOGGER = logging.getLogger(__name__)
LOGGER.setLevel(logging.INFO)

VALID_SEMESTERS = {"1", "2", "3", "SUMMER", "OTHER"}
PUBLIC_VISIBILITY = "PUBLIC"
ACTIVE_STATUS = "ACTIVE"

JSON_HEADERS = {
    "content-type": "application/json; charset=utf-8",
    "access-control-allow-origin": "*",
}


class MasterDataQueryError(Exception):
    """Validation error returned to public API callers."""

    def __init__(self, field: str, message: str) -> None:
        super().__init__(message)
        self.code = "INVALID_QUERY"
        self.status = 400
        self.details = {"field": field}


class MasterDataRouteError(Exception):
    """Route/method error returned to public API callers."""

    def __init__(self, status: int, code: str, message: str) -> None:
        super().__init__(message)
        self.status = status
        self.code = code


def _json_response(status_code: int, body: dict[str, Any]) -> dict[str, Any]:
    return {
        "statusCode": status_code,
        "headers": JSON_HEADERS,
        "body": json.dumps(body, ensure_ascii=False, separators=(",", ":")),
    }


def _error_response(error: Exception) -> dict[str, Any]:
    if isinstance(error, MasterDataQueryError):
        return _json_response(
            error.status,
            {
                "error": {
                    "code": error.code,
                    "message": str(error),
                    "details": error.details,
                }
            },
        )

    if isinstance(error, MasterDataRouteError):
        return _json_response(
            error.status,
            {
                "error": {
                    "code": error.code,
                    "message": str(error),
                }
            },
        )

    LOGGER.exception("Unhandled V2 master data API error")
    return _json_response(
        500,
        {
            "error": {
                "code": "INTERNAL_ERROR",
                "message": "Unable to load V2 master data",
            }
        },
    )


def _envelope(items: list[dict[str, Any]]) -> dict[str, Any]:
    return {"items": items, "meta": {"count": len(items)}}


def _query_params(event: dict[str, Any]) -> dict[str, str]:
    params = event.get("queryStringParameters") or {}
    return {str(key): str(value) for key, value in params.items() if value is not None}


def _optional_param(params: dict[str, str], key: str) -> str | None:
    value = params.get(key)
    if value is None:
        return None
    stripped = value.strip()
    return stripped or None


def _method(event: dict[str, Any]) -> str:
    request_context = event.get("requestContext") or {}
    http = request_context.get("http") or {}
    return str(http.get("method") or event.get("httpMethod") or "GET").upper()


def _path(event: dict[str, Any]) -> str:
    request_context = event.get("requestContext") or {}
    http = request_context.get("http") or {}
    raw_path = event.get("rawPath") or http.get("path") or event.get("path") or "/"
    path = str(raw_path).rstrip("/")
    return path or "/"


def _parse_year(params: dict[str, str]) -> int | None:
    year = _optional_param(params, "year")
    if year is None:
        return None
    if not year.isdigit() or len(year) != 4:
        raise MasterDataQueryError("year", "year must be a 4-digit academic year")
    return int(year)


def _parse_semester(params: dict[str, str]) -> str | None:
    semester = _optional_param(params, "semester")
    if semester is None:
        return None
    if semester not in VALID_SEMESTERS:
        raise MasterDataQueryError("semester", "semester must be one of 1, 2, 3, SUMMER, OTHER")
    return semester


def _parse_active(params: dict[str, str]) -> bool:
    active = _optional_param(params, "active")
    if active is None:
        return True
    if active == "true":
        return True
    if active == "false":
        return False
    raise MasterDataQueryError("active", "active must be true or false")


def _parse_public_status(params: dict[str, str]) -> str:
    status = _optional_param(params, "status") or ACTIVE_STATUS
    if status != ACTIVE_STATUS:
        raise MasterDataQueryError("status", "status must be ACTIVE for public master data endpoints")
    return status


def _parse_public_visibility(params: dict[str, str]) -> str:
    visibility = _optional_param(params, "visibility") or PUBLIC_VISIBILITY
    if visibility != PUBLIC_VISIBILITY:
        raise MasterDataQueryError(
            "visibility", "visibility must be PUBLIC for public master data endpoints"
        )
    return visibility


def _public_faculty(row: dict[str, Any]) -> dict[str, Any]:
    return {
        "id": row.get("id"),
        "public_slug": row.get("public_slug"),
        "name_th": row.get("name_th"),
        "name_en": row.get("name_en"),
        "academic_position": row.get("academic_position"),
        "department": row.get("department"),
        "profile_image_url": row.get("profile_image_url"),
        "profile_image_alt": row.get("profile_image_alt"),
    }


class MasterDataRepository:
    """Storage boundary used by the Lambda handler."""

    def list_academic_periods(self, year: int | None, semester: str | None) -> list[dict[str, Any]]:
        raise NotImplementedError

    def list_evaluation_periods(self, academic_period_id: str | None) -> list[dict[str, Any]]:
        raise NotImplementedError

    def list_work_categories(self, active: bool) -> list[dict[str, Any]]:
        raise NotImplementedError

    def work_category_exists(self, category_code: str) -> bool:
        raise NotImplementedError

    def list_work_types(self, active: bool, category_code: str | None) -> list[dict[str, Any]]:
        raise NotImplementedError

    def list_faculties(self, query: str | None, status: str, visibility: str) -> list[dict[str, Any]]:
        raise NotImplementedError


@dataclass
class DataApiMasterDataRepository(MasterDataRepository):
    resource_arn: str
    secret_arn: str
    database: str
    client: Any | None = None

    def __post_init__(self) -> None:
        if self.client is None:
            import boto3

            self.client = boto3.client("rds-data")

    def execute(self, sql: str, params: dict[str, Any] | None = None) -> list[dict[str, Any]]:
        request: dict[str, Any] = {
            "resourceArn": self.resource_arn,
            "secretArn": self.secret_arn,
            "database": self.database,
            "sql": sql,
        }
        if params:
            request["parameters"] = [_data_api_param(name, value) for name, value in params.items()]

        response = self.client.execute_statement(**request)
        return _records_to_rows(response.get("columnMetadata", []), response.get("records", []))

    def list_academic_periods(self, year: int | None, semester: str | None) -> list[dict[str, Any]]:
        where: list[str] = []
        params: dict[str, Any] = {}
        if year is not None:
            where.append("academic_year = :year")
            params["year"] = year
        if semester is not None:
            where.append("semester = :semester")
            params["semester"] = semester

        where_clause = f"WHERE {' AND '.join(where)}" if where else ""
        return self.execute(
            f"""
            SELECT id, academic_year, semester, label,
                   start_date::text AS start_date,
                   end_date::text AS end_date
            FROM academic_period
            {where_clause}
            ORDER BY academic_year DESC,
              CASE semester
                WHEN '3' THEN 5
                WHEN '2' THEN 4
                WHEN '1' THEN 3
                WHEN 'SUMMER' THEN 2
                WHEN 'OTHER' THEN 1
                ELSE 0
              END DESC
            """,
            params,
        )

    def list_evaluation_periods(self, academic_period_id: str | None) -> list[dict[str, Any]]:
        where = ""
        params: dict[str, Any] = {}
        if academic_period_id:
            where = "WHERE academic_period_id = :academic_period_id"
            params["academic_period_id"] = academic_period_id

        return self.execute(
            f"""
            SELECT id, code, label,
                   start_date::text AS start_date,
                   end_date::text AS end_date,
                   academic_period_id
            FROM evaluation_period
            {where}
            ORDER BY start_date DESC, code ASC
            """,
            params,
        )

    def list_work_categories(self, active: bool) -> list[dict[str, Any]]:
        return self.execute(
            """
            SELECT code, label_th, label_en, description, display_order, is_active
            FROM work_category
            WHERE is_active = :active
            ORDER BY display_order ASC, code ASC
            """,
            {"active": active},
        )

    def work_category_exists(self, category_code: str) -> bool:
        rows = self.execute(
            """
            SELECT code
            FROM work_category
            WHERE code = :category_code
            LIMIT 1
            """,
            {"category_code": category_code},
        )
        return bool(rows)

    def list_work_types(self, active: bool, category_code: str | None) -> list[dict[str, Any]]:
        where = ["wt.is_active = :active"]
        params: dict[str, Any] = {"active": active}
        if category_code:
            where.append("wt.category_code = :category_code")
            params["category_code"] = category_code

        return self.execute(
            f"""
            SELECT wt.code, wt.category_code, wt.label_th, wt.label_en,
                   wt.description, wt.default_visibility, wt.display_order, wt.is_active
            FROM work_type wt
            JOIN work_category wc ON wc.code = wt.category_code
            WHERE {' AND '.join(where)}
            ORDER BY wc.display_order ASC, wt.display_order ASC, wt.code ASC
            """,
            params,
        )

    def list_faculties(self, query: str | None, status: str, visibility: str) -> list[dict[str, Any]]:
        where = ["status = :status", "visibility = :visibility"]
        params: dict[str, Any] = {"status": status, "visibility": visibility}

        if query:
            where.append(
                """(
                    public_slug ILIKE :query
                    OR name_th ILIKE :query
                    OR name_en ILIKE :query
                    OR academic_position ILIKE :query
                    OR department ILIKE :query
                )"""
            )
            params["query"] = f"%{query}%"

        return self.execute(
            f"""
            SELECT id, public_slug, name_th, name_en, academic_position, department,
                   profile_image_url, profile_image_alt
            FROM faculty
            WHERE {' AND '.join(where)}
            ORDER BY COALESCE(NULLIF(name_th, ''), NULLIF(name_en, ''), public_slug) ASC
            """,
            params,
        )


def _data_api_param(name: str, value: Any) -> dict[str, Any]:
    field: dict[str, Any]
    if value is None:
        field = {"isNull": True}
    elif isinstance(value, bool):
        field = {"booleanValue": value}
    elif isinstance(value, int):
        field = {"longValue": value}
    elif isinstance(value, float):
        field = {"doubleValue": value}
    else:
        field = {"stringValue": str(value)}
    return {"name": name, "value": field}


def _field_value(field: dict[str, Any]) -> Any:
    if field.get("isNull"):
        return None
    for key in ("stringValue", "longValue", "doubleValue", "booleanValue"):
        if key in field:
            return field[key]
    if "blobValue" in field:
        return field["blobValue"]
    return None


def _records_to_rows(
    column_metadata: Iterable[dict[str, Any]], records: Iterable[list[dict[str, Any]]]
) -> list[dict[str, Any]]:
    columns = [column.get("label") or column.get("name") for column in column_metadata]
    rows: list[dict[str, Any]] = []
    for record in records:
        rows.append({str(column): _field_value(field) for column, field in zip(columns, record)})
    return rows


class MasterDataService:
    def __init__(self, repository: MasterDataRepository) -> None:
        self.repository = repository

    def list_academic_periods(self, params: dict[str, str]) -> dict[str, Any]:
        items = self.repository.list_academic_periods(_parse_year(params), _parse_semester(params))
        return _envelope(items)

    def list_evaluation_periods(self, params: dict[str, str]) -> dict[str, Any]:
        academic_period_id = _optional_param(params, "academic_period_id")
        return _envelope(self.repository.list_evaluation_periods(academic_period_id))

    def list_work_categories(self, params: dict[str, str]) -> dict[str, Any]:
        return _envelope(self.repository.list_work_categories(_parse_active(params)))

    def list_work_types(self, params: dict[str, str]) -> dict[str, Any]:
        active = _parse_active(params)
        category = _optional_param(params, "category")
        if category and not self.repository.work_category_exists(category):
            raise MasterDataQueryError("category", "category must be a known work category code")
        return _envelope(self.repository.list_work_types(active, category))

    def list_faculties(self, params: dict[str, str]) -> dict[str, Any]:
        query = _optional_param(params, "q")
        status = _parse_public_status(params)
        visibility = _parse_public_visibility(params)
        items = [_public_faculty(row) for row in self.repository.list_faculties(query, status, visibility)]
        return _envelope(items)


_DEFAULT_REPOSITORY: DataApiMasterDataRepository | None = None


def _default_repository() -> DataApiMasterDataRepository:
    global _DEFAULT_REPOSITORY
    if _DEFAULT_REPOSITORY is None:
        resource_arn = os.environ["DB_CLUSTER_ARN"]
        secret_arn = os.environ["DB_SECRET_ARN"]
        database = os.environ["DB_NAME"]
        _DEFAULT_REPOSITORY = DataApiMasterDataRepository(resource_arn, secret_arn, database)
    return _DEFAULT_REPOSITORY


def handle_request(event: dict[str, Any], repository: MasterDataRepository | None = None) -> dict[str, Any]:
    try:
        if _method(event) != "GET":
            raise MasterDataRouteError(405, "METHOD_NOT_ALLOWED", "Only GET is supported")

        params = _query_params(event)
        service = MasterDataService(repository or _default_repository())
        path = _path(event)

        routes = {
            "/api/v2/academic-periods": service.list_academic_periods,
            "/api/v2/evaluation-periods": service.list_evaluation_periods,
            "/api/v2/work-categories": service.list_work_categories,
            "/api/v2/work-types": service.list_work_types,
            "/api/v2/faculties": service.list_faculties,
        }
        route = routes.get(path)
        if route is None:
            raise MasterDataRouteError(404, "NOT_FOUND", "V2 master data route not found")

        body = route(params)
        LOGGER.info(
            "v2_master_data_request path=%s status=200 count=%s",
            path,
            body.get("meta", {}).get("count"),
        )
        return _json_response(200, body)
    except Exception as error:
        return _error_response(error)


def handler(event: dict[str, Any], context: Any) -> dict[str, Any]:
    return handle_request(event)

File: frontend/lib/v2/master-data.mjs
import fs from 'node:fs';
import path from 'node:path';

const FIXTURE_FILES = {
  academicPeriods: 'academic-periods.json',
  evaluationPeriods: 'evaluation-periods.json',
  workCategories: 'work-categories.json',
  workTypes: 'work-types.json',
  faculties: 'faculties.json',
};

const VALID_SEMESTERS = new Set(['1', '2', '3', 'SUMMER', 'OTHER']);
const SEMESTER_ORDER = new Map([
  ['3', 5],
  ['2', 4],
  ['1', 3],
  ['SUMMER', 2],
  ['OTHER', 1],
]);

const PUBLIC_VISIBILITY = 'PUBLIC';
const ACTIVE_STATUS = 'ACTIVE';

export class MasterDataQueryError extends Error {
  constructor(field, message) {
    super(message);
    this.name = 'MasterDataQueryError';
    this.code = 'INVALID_QUERY';
    this.status = 400;
    this.details = { field };
  }
}

function getFixtureDirectory() {
  if (process.env.V2_MASTER_DATA_FIXTURE_DIR) {
    return path.resolve(process.env.V2_MASTER_DATA_FIXTURE_DIR);
  }

  const candidates = [
    path.resolve(process.cwd(), 'data', 'v2', 'fixtures'),
    path.resolve(process.cwd(), '..', 'data', 'v2', 'fixtures'),
  ];

  const fixtureDirectory = candidates.find((candidate) => fs.existsSync(candidate));
  return fixtureDirectory ?? candidates[1];
}

function readFixture(fileName) {
  const filePath = path.join(getFixtureDirectory(), fileName);
  return JSON.parse(fs.readFileSync(filePath, 'utf8'));
}

export function loadFixtureRepository() {
  return {
    academicPeriods: readFixture(FIXTURE_FILES.academicPeriods),
    evaluationPeriods: readFixture(FIXTURE_FILES.evaluationPeriods),
    workCategories: readFixture(FIXTURE_FILES.workCategories),
    workTypes: readFixture(FIXTURE_FILES.workTypes),
    faculties: readFixture(FIXTURE_FILES.faculties),
  };
}

function asSearchParams(searchParams) {
  if (searchParams instanceof URLSearchParams) return searchParams;
  return new URLSearchParams(searchParams ?? '');
}

function optionalParam(searchParams, key) {
  const value = asSearchParams(searchParams).get(key);
  const normalized = value?.trim();
  return normalized ? normalized : undefined;
}

function makeEnvelope(items) {
  return {
    items,
    meta: {
      count: items.length,
    },
  };
}

function parseAcademicYear(searchParams) {
  const year = optionalParam(searchParams, 'year');
  if (!year) return undefined;
  if (!/^\d{4}$/.test(year)) {
    throw new MasterDataQueryError('year', 'year must be a 4-digit academic year');
  }
  return Number(year);
}

function parseSemester(searchParams) {
  const semester = optionalParam(searchParams, 'semester');
  if (!semester) return undefined;
  if (!VALID_SEMESTERS.has(semester)) {
    throw new MasterDataQueryError('semester', 'semester must be one of 1, 2, 3, SUMMER, OTHER');
  }
  return semester;
}

function parseActive(searchParams) {
  const active = optionalParam(searchParams, 'active');
  if (active === undefined) return true;
  if (active === 'true') return true;
  if (active === 'false') return false;
  throw new MasterDataQueryError('active', 'active must be true or false');
}

function parsePublicStatus(searchParams) {
  const status = optionalParam(searchParams, 'status') ?? ACTIVE_STATUS;
  if (status !== ACTIVE_STATUS) {
    throw new MasterDataQueryError('status', 'status must be ACTIVE for public master data endpoints');
  }
  return status;
}

function parsePublicVisibility(searchParams) {
  const visibility = optionalParam(searchParams, 'visibility') ?? PUBLIC_VISIBILITY;
  if (visibility !== PUBLIC_VISIBILITY) {
    throw new MasterDataQueryError(
      'visibility',
      'visibility must be PUBLIC for public master data endpoints'
    );
  }
  return visibility;
}

export function mapAcademicPeriod(row) {
  return {
    id: row.id,
    academic_year: row.academic_year,
    semester: row.semester,
    label: row.label,
    start_date: row.start_date ?? null,
    end_date: row.end_date ?? null,
  };
}

export function mapEvaluationPeriod(row) {
  return {
    id: row.id,
    code: row.code,
    label: row.label,
    start_date: row.start_date,
    end_date: row.end_date,
    academic_period_id: row.academic_period_id ?? null,
  };
}

export function mapWorkCategory(row) {
  return {
    code: row.code,
    label_th: row.label_th,
    label_en: row.label_en,
    description: row.description ?? null,
    display_order: row.display_order,
    is_active: Boolean(row.is_active),
  };
}

export function mapWorkType(row) {
  return {
    code: row.code,
    category_code: row.category_code,
    label_th: row.label_th,
    label_en: row.label_en,
    description: row.description ?? null,
    default_visibility: row.default_visibility,
    display_order: row.display_order,
    is_active: Boolean(row.is_active),
  };
}

export function mapFacultyOption(row) {
  return {
    id: row.id,
    public_slug: row.public_slug,
    name_th: row.name_th ?? null,
    name_en: row.name_en ?? null,
    academic_position: row.academic_position ?? null,
    department: row.department ?? null,
    profile_image_url: row.profile_image_url ?? null,
    profile_image_alt: row.profile_image_alt ?? null,
  };
}

function sortAcademicPeriods(left, right) {
  if (right.academic_year !== left.academic_year) {
    return right.academic_year - left.academic_year;
  }
  return (SEMESTER_ORDER.get(right.semester) ?? 0) - (SEMESTER_ORDER.get(left.semester) ?? 0);
}

function sortText(left, right) {
  return left.localeCompare(right, 'th');
}

export function listAcademicPeriods(searchParams) {
  const year = parseAcademicYear(searchParams);
  const semester = parseSemester(searchParams);
  const repository = loadFixtureRepository();

  const items = repository.academicPeriods
    .map(mapAcademicPeriod)
    .filter((period) => year === undefined || period.academic_year === year)
    .filter((period) => semester === undefined || period.semester === semester)
    .sort(sortAcademicPeriods);

  return makeEnvelope(items);
}

export function listEvaluationPeriods(searchParams) {
  const academicPeriodId = optionalParam(searchParams, 'academic_period_id');
  const repository = loadFixtureRepository();

  const items = repository.evaluationPeriods
    .map(mapEvaluationPeriod)
    .filter((period) => !academicPeriodId || period.academic_period_id === academicPeriodId)
    .sort((left, right) => {
      const dateOrder = right.start_date.localeCompare(left.start_date);
      return dateOrder || sortText(left.code, right.code);
    });

  return makeEnvelope(items);
}

export function listWorkCategories(searchParams) {
  const active = parseActive(searchParams);
  const repository = loadFixtureRepository();

  const items = repository.workCategories
    .map(mapWorkCategory)
    .filter((category) => category.is_active === active)
    .sort((left, right) => left.display_order - right.display_order || sortText(left.code, right.code));

  return makeEnvelope(items);
}

export function listWorkTypes(searchParams) {
  const active = parseActive(searchParams);
  const category = optionalParam(searchParams, 'category');
  const repository = loadFixtureRepository();
  const knownCategories = new Set(repository.workCategories.map((item) => item.code));

  if (category && !knownCategories.has(category)) {
    throw new MasterDataQueryError('category', 'category must be a known work category code');
  }

  const categoryOrder = new Map(
    repository.workCategories.map((item) => [item.code, item.display_order ?? 0])
  );

  const items = repository.workTypes
    .map(mapWorkType)
    .filter((type) => type.is_active === active)
    .filter((type) => !category || type.category_code === category)
    .sort((left, right) => {
      const categorySort =
        (categoryOrder.get(left.category_code) ?? 0) - (categoryOrder.get(right.category_code) ?? 0);
      return categorySort || left.display_order - right.display_order || sortText(left.code, right.code);
    });

  return makeEnvelope(items);
}

export function listFaculties(searchParams) {
  parsePublicStatus(searchParams);
  parsePublicVisibility(searchParams);

  const query = optionalParam(searchParams, 'q')?.toLocaleLowerCase('th');
  const repository = loadFixtureRepository();

  const items = repository.faculties
    .filter((faculty) => faculty.status === ACTIVE_STATUS)
    .filter((faculty) => faculty.visibility === PUBLIC_VISIBILITY)
    .filter((faculty) => {
      if (!query) return true;
      return [
        faculty.public_slug,
        faculty.name_th,
        faculty.name_en,
        faculty.academic_position,
        faculty.department,
      ]
        .filter(Boolean)
        .some((value) => String(value).toLocaleLowerCase('th').includes(query));
    })
    .map(mapFacultyOption)
    .sort((left, right) => {
      const leftName = left.name_th ?? left.name_en ?? left.public_slug;
      const rightName = right.name_th ?? right.name_en ?? right.public_slug;
      return sortText(leftName, rightName);
    });

  return makeEnvelope(items);
}

export function formatMasterDataError(error) {
  if (error instanceof MasterDataQueryError) {
    return {
      status: error.status,
      body: {
        error: {
          code: error.code,
          message: error.message,
          details: error.details,
        },
      },
    };
  }

  return {
    status: 500,
    body: {
      error: {
        code: 'INTERNAL_ERROR',
        message: 'Unable to load V2 master data',
      },
    },
  };
}

const masterData = {
  MasterDataQueryError,
  FIXTURE_FILES,
  formatMasterDataError,
  listAcademicPeriods,
  listEvaluationPeriods,
  listFaculties,
  listWorkCategories,
  listWorkTypes,
  loadFixtureRepository,
  mapAcademicPeriod,
  mapEvaluationPeriod,
  mapFacultyOption,
  mapWorkCategory,
  mapWorkType,
};

export { FIXTURE_FILES };

export default masterData;
